monitor/diskspace.go

```diff
@@ -0,0 +1,112 @@
+package monitor
+
+import (
+	"errors"
+	"sync"
+	"time"
+
+	"github.com/shirou/gopsutil/v3/disk"
+	log "github.com/sirupsen/logrus"
+
+	"github.com/futurehomeno/cliffhanger/core"
+)
+
+// DiskSpace represents a disk space monitor.
+type DiskSpace interface {
+	core.Service
+
+	DiskFull() bool
+}
+
+type diskSpace struct {
+	interval     time.Duration
+	limitPercent float64
+	used         float64
+
+	closeCh   chan struct{}
+	lock      *sync.Mutex
+	waitGroup *sync.WaitGroup
+}
+
+// NewDiskSpace creates a new instance of disk space monitor.
+func NewDiskSpace(interval time.Duration, limitPercent float64) DiskSpace {
+	return &diskSpace{
+		interval:     interval,
+		limitPercent: limitPercent,
+		lock:         &sync.Mutex{},
+		waitGroup:    &sync.WaitGroup{},
+	}
+}
+
+// DiskFull returns true if the disk space is on limit.
+func (d *diskSpace) DiskFull() bool {
+	if d.used == 0 {
+		d.checkSpace()
+	}
+
+	return d.used >= d.limitPercent
```

Comment:
Add read lock here. You have race condition between line 47 and 95->111.

The cleanest way would be to use RWMutex dedicated to writing/reading d.used instead of main lock for everything, but it's a minor detail.